From d19b1b91ff32f7e22fb31eb78c693717538b008e Mon Sep 17 00:00:00 2001
From: Shengjiu Wang <shengjiu.wang@freescale.com>
Date: Sat, 14 May 2016 14:50:22 +0800
Subject: [PATCH 1061/5242] MLK-12794: ASoC: fsl_asrc: fix underrun issue when
 convert 192k to 96kHz.

commit  d055634c39d12e78730657bfbc03d3a2721848e1 from
https://source.codeaurora.org/external/imx/linux-imx.git

The maximum divider of asrc clock is 1024, but there is no judgement
for this limitaion in driver, which may cause the divider setting not
correct.

When IDEAL_RATIO_RATE 200kHZ, the cost time of conversion from 192kHz
to 96kHz is 24ms every 1024 sample, but these sample's playback time
is 1024/96=11ms, so there will be underrun. So need to enlarge this RATE.

Signed-off-by: Shengjiu Wang <shengjiu.wang@freescale.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 sound/soc/fsl/fsl_asrc.c |    6 ++++++
 sound/soc/fsl/fsl_asrc.h |    4 ++--
 2 files changed, 8 insertions(+), 2 deletions(-)

diff --git a/sound/soc/fsl/fsl_asrc.c b/sound/soc/fsl/fsl_asrc.c
index 340aebb..1dc9b6c 100644
--- a/sound/soc/fsl/fsl_asrc.c
+++ b/sound/soc/fsl/fsl_asrc.c
@@ -356,6 +356,12 @@ static int fsl_asrc_config_pair(struct fsl_asrc_pair *pair, bool p2p_in, bool p2
 		return -EINVAL;
 	}
 
+	if (div[IN] > 1024)
+		div[IN] = 1024;
+
+	if (div[OUT] > 1024)
+		div[OUT] = 1024;
+
 	/* Set the channel number */
 	channels = config->channel_num;
 
diff --git a/sound/soc/fsl/fsl_asrc.h b/sound/soc/fsl/fsl_asrc.h
index c67057a..5b353f9 100644
--- a/sound/soc/fsl/fsl_asrc.h
+++ b/sound/soc/fsl/fsl_asrc.h
@@ -1,7 +1,7 @@
 /*
  * fsl_asrc.h - Freescale ASRC ALSA SoC header file
  *
- * Copyright (C) 2014 Freescale Semiconductor, Inc.
+ * Copyright (C) 2014-2016 Freescale Semiconductor, Inc.
  *
  * Author: Nicolin Chen <nicoleotsuka@gmail.com>
  *
@@ -28,7 +28,7 @@
 #define ASRC_OUTPUT_LAST_SAMPLE_MAX	32
 #define ASRC_OUTPUT_LAST_SAMPLE		16
 
-#define IDEAL_RATIO_RATE		200000
+#define IDEAL_RATIO_RATE		1000000
 
 #define REG_ASRCTR			0x00
 #define REG_ASRIER			0x04
-- 
1.7.9.5

